File: src/cpp/linear_search.cpp
#include <iostream>
#include <vector>


template <typename T>
int linear_search(std::vector<T> seq, int query) {
  for (size_t i = 0; i < seq.size(); ++i) {
    if (seq[i] == query) {
      return i;
    }
  }
  return -1;
}


void run_tests() {
  std::vector<int> seq = {2, 3, 4, 5, 7, 9, 12, 16, 21, 28, 37, 49, 65};

  std::cout << "Search result : " << linear_search(seq, 7) << "\n";
  std::cout << "Missing query : " << linear_search(seq, 6) << "\n";
  std::cout << "Empty sequence: " << linear_search(std::vector<int> {}, 7)
    << std::endl;
}


int main() {
  run_tests();
  return 0;
}

File: src/py3/linear_search.py
from typing import Sequence


def linear_search(sequence: Sequence[int], query: int) -> int:
    for i, val in enumerate(sequence):
        if val == query:
            return i

    return -1


if __name__ == "__main__":
    seq = [2, 3, 4, 5, 7, 9, 12, 16, 21, 28, 37, 49, 65]
    print(f"Search result : {linear_search(seq, 7)}")
    print(f"Missing query : {linear_search(seq, 6)}")
    print(f"Empty sequence: {linear_search([], 7)}")
